cocotb/run.py

```diff
@@ -0,0 +1,184 @@
+
+    python_lib = sysconfig.get_config_var("LDLIBRARY")
+    python_lib_link = os.path.splitext(python_lib)[0][3:]
+
+    dist = Distribution()
+    dist.parse_config_files()
+
+    libcocotbutils = Extension(
+        "libcocotbutils",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/utils/cocotb_utils.c"],
+    )
+
+    lib_path, ext_name = _build_lib(libcocotbutils, dist)
+
+    libgpilog = Extension(
+        "libgpilog",
+        include_dirs=[share_dir + "/include"],
+        libraries=[python_lib_link, "pthread", "dl", "util", "rt", "m", "cocotbutils"],
+        library_dirs=[lib_path],
+        sources=[share_dir + "/lib/gpi_log/gpi_logging.c"],
+    )
+
+    _build_lib(libgpilog, dist)
+
+    libcocotb = Extension(
+        "libcocotb",
+        define_macros=[("PYTHON_SO_LIB", python_lib)],
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/embed/gpi_embed.c"],
+    )
+
+    _build_lib(libcocotb, dist)
+
+    libgpi = Extension(
+        "libgpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["cocotbutils", "gpilog", "cocotb", "stdc++"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/gpi/GpiCbHdl.cpp",
+            share_dir + "/lib/gpi/GpiCommon.cpp",
+        ],
+    )
+
+    _build_lib(libgpi, dist)
+
+    libsim = Extension(
+        "simulator",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/simulator/simulatormodule.c"],
+    )
+
+    _build_lib(libsim, dist)
+
+    libvpi = Extension(
+        "libvpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["gpi", "gpilog"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/vpi/VpiImpl.cpp",
+            share_dir + "/lib/vpi/VpiCbHdl.cpp",
+        ],
+    )
+
+    _build_lib(libvpi, dist)
+
+    _symlink_force(
+        os.path.join(lib_path, "libvpi." + ext_name),
+        os.path.join(lib_path, "gpivpi.vpl"),
+    )
+    _symlink_force(
+        os.path.join(lib_path, "libvpi." + ext_name),
+        os.path.join(lib_path, "cocotb.vpl"),
+    )
+
+    return lib_path
+
+
+def Run(sources, toplevel, module=None):
+
+    libs_dir = build_libs()
+
+    previous_frame = inspect.currentframe().f_back
+    (run_module_filename, _, _, _, _) = inspect.getframeinfo(previous_frame)
+
+    run_dir_name = os.path.dirname(run_module_filename)
+    run_module_name = os.path.splitext(os.path.split(run_module_filename)[-1])[0]
+
+    if module is None:
+        module = run_module_name
+
+    my_env = os.environ
+    my_env["LD_LIBRARY_PATH"] = libs_dir + ":" + sysconfig.get_config_var("LIBDIR")
+
+    python_path = ":".join(sys.path)
+    my_env["PYTHONPATH"] = python_path + ":" + libs_dir
+    my_env["TOPLEVEL"] = toplevel
+    my_env["TOPLEVEL_LANG"] = "verilog"
+    my_env["COCOTB_SIM"] = "1"
+    my_env["MODULE"] = module
```

Comment:
A neat trick would be to pass the test functions between embedded and non-embedded python using pickle, like how `multiprocessing` does.

Reply:
Can be done but I guess we keep Makefile compatibility for now?